test/unit/range/iterator_test_template.hpp

```diff
@@ -70,6 +70,29 @@ void expect_iter_equal(it_t && it, rng_it_t && rng_it)
     expect_iter_value_equal<T>(*it, *rng_it);
 }
 
+// std c++20 input iterator aren't required to have an operator==(iterator_t, iterator_t), but if they have one we
+// test the semantic
+template <typename type_param_t>
+SEQAN3_CONCEPT iterator_is_equality_comparable =
+    std::derived_from<typename iterator_fixture<type_param_t>::iterator_tag, std::forward_iterator_tag> ||
+#if SEQAN3_WORKAROUND_GCC7_AND_8_CONCEPT_ISSUES
+    requires (decltype(std::ranges::begin(std::declval<iterator_fixture<type_param_t> &>().test_range)) & it)
+{
+    { it == it };
+};
+#else // ^^^ workaround / no workaround vvv
+    requires (iterator_fixture<type_param_t> & fixture)
+{
+    typename std::ranges::iterator_t<decltype(fixture.test_range)>;
+
+    requires requires(std::ranges::iterator_t<decltype(fixture.test_range)> & it)
+    {
+        // we don't assume anything about the return type, this will be done in the tests
+        { it == it };
```

Comment:
This isn't a complete concept, it just checks that one operator is defined to enable certain tests. The test case will then test all possible variants of it, and it also should fail to compile if for example the `operator!=` is missing.

As a side note, in C++20 an `operator!=` will always be defined if an `operator==` exists.

Reply:
So there is the line of code:
```
    if constexpr(iterator_is_equality_comparable<TypeParam>)
    {
        EXPECT_TRUE(std::ranges::begin(this->test_range) == std::ranges::begin(this->test_range));
        EXPECT_FALSE(std::ranges::begin(this->test_range) != std::ranges::begin(this->test_range));
    }
    ```
    This requires `operator !=` to exist. If I understand correct, C++20 will default define `operator!=` when `operator==` exists. But in theory I could define `operator==` and delete `operator !=`, which means the concept would be valid, but the test would fail to compile.

Reply:
> This requires `operator !=` to exist. If I understand correct, C++20 will default define `operator!=` when `operator==` exists. But in theory I could define `operator==` and delete `operator !=`, which means the concept would be valid, but the test would fail to compile.

True, but I want that behaviour. The concept here is kind of misused as a syntax checker: If iterator type has an `operator==` then do the following test. And in the context of an iterator it is weird to have an equal operator, bot not an unequal operator.